feature: Support new release workflow and set User-Agent header with SDK version

kevihan-te opened the pull request:
- Add a new workflow which can be triggered manually to release next version of the SDK 
  - The target version will be stored in the `go.mod` file
  - The version number will be used as the git tag

- Support parse the go.mod to extract the SDK version number in the runtime
- Use the SDK version to build the `User-Agent` header

kevihan-te committed:
+ add a manual workflow to release sdk and update go.mod file
+ update the configuration field to dynamically parse sdk version from the go.mod

kevihan-te committed:
add debug command

kevihan-te committed:
format sdk

kevihan-te commented on client/configuration.go:
```diff
@@ -2,13 +2,18 @@ package client
 type Configuration struct {
-	UserAgent  string `json:"userAgent,omitempty"`
-	Debug      bool   `json:"debug,omitempty"`
+	userAgent  string
```
Make it private to make sure users can not modify it.

jabrowne-te commented on .github/workflows/release-sdk.yml:
```diff
@@ -0,0 +1,32 @@
+
+      - name: Upgrade the version number in the go.mod
+        run: |
+          echo "Releasing the target version: ${{ github.event.inputs.releaseVersion }}"
+          sed -i "s|^\(module github.com/thousandeyes/thousandeyes-sdk-go/\).*|\1${{ github.event.inputs.releaseVersion }}|" go.mod
```
Have I understood correctly that this will replace the string `module github.com/thousandeyes/thousandeyes-sdk-go/v3` with `module github.com/thousandeyes/thousandeyes-sdk-go/v3.1.0`, for example?

If so, I'm not sure we can do that, the syntax of the `module` declaration in the `go.mod` file is pretty unambiguous:

> The module's module path, usually the repository location from which the module can be downloaded by Go tools. For module versions v2 and later, this value must end with the major version number, such as /v2.

[source](https://go.dev/doc/modules/gomod-ref#module)

kevihan-te replied:
Thanks Jack. It should be an explicit standard in go document. I updated it. Pls take a look again.

brumarqu-te commented on client/configuration.go:
```diff
@@ -18,13 +23,28 @@ type Configuration struct {
 	return cfg
 }
 
+func getCurrentModuleVersion() string {
+	data, err := os.ReadFile("go.mod")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	modFile, err := modfile.Parse("go.mod", data, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	parts := strings.Split(modFile.Module.Mod.Path, "/")
+	return parts[len(parts)-1]
+}
+
```
instead of having to parse a file to read the version can we use a const in some `version.go` file and read it directly from there?

kevihan-te replied:
Thanks Bruno. Because of the problem raised by Jack, this method will not work. I introduced a new variable to stand for the sdk version and used the workflow to update it each time.

kevihan-te committed:
+ use major version in module version
+ use init method to init and reuse version number
+ support customize user-agent

kevihan-te committed:
+ directly update the version var in the workflow
+ only keep major version in the go.mod version

jabrowne-te commented on client/configuration.go:
```diff
@@ -18,13 +27,20 @@ type Configuration struct {
 	return cfg
 }
 
+func (c *Configuration) GenerateUserAgent() string {
```
nit: maybe `BuildUserAgent` would be a better name here? I don't think this is really "generating" anything, it's just constructing the User Agent string from pre-defined parts

kevihan-te replied:
Good idea. I updated. Thanks Jack.

mpragosa-te commented on .github/workflows/release-sdk.yml:
```diff
@@ -0,0 +1,35 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
```
Checkout's latest version is `actions/checkout@v4`. Unless there's a strong reason to use v2, we should use v4 instead.

kevihan-te replied:
Updated. Thanks Miguel.

mpragosa-te commented on .github/workflows/release-sdk.yml:
```diff
@@ -0,0 +1,35 @@
+    inputs:
+      releaseVersion:
+        description: 'The target version you want to update and release'
+        required: true
```
Don't we need to set a `type` too? I was under the impression that it's a mandatory field, but happy to keep things as is in case it's not.

mpragosa-te replied:
This comment also applies to https://github.com/thousandeyes/thousandeyes-sdk-java/pull/106 and https://github.com/thousandeyes/thousandeyes-sdk-python/pull/94

kevihan-te replied:
It can work without `type` field. But I agree it's a good idea to include this field which make everything more explicit

mpragosa-te commented on .github/workflows/release-sdk.yml:
```diff
@@ -0,0 +1,35 @@
+      - name: Upgrade the version number in the configuration.go and go.mod files
+        run: |
+          echo "Releasing the target version: ${{ github.event.inputs.releaseVersion }}"
+          major_version=$(echo "${{ github.event.inputs.releaseVersion }}" | cut -d '.' -f 1)
```
nit: Looking at https://github.com/thousandeyes/open-api-specifications/blob/main/.github/workflows/post-release.yaml, I believe `inputs.releaseVersion` would be enough

mpragosa-te replied:
This comment also applies to https://github.com/thousandeyes/thousandeyes-sdk-java/pull/106 and https://github.com/thousandeyes/thousandeyes-sdk-python/pull/94

kevihan-te replied:
Updated. Thanks Miguel.

kevihan-te committed:
+ rename buildUserAgent
+ use latest version of git checkout
+ use simplified name to refer git workflow input

mpragosa-te commented on .github/workflows/release-sdk.yml:
```diff
@@ -0,0 +1,36 @@
+name: Release SDK
```
Sorry Kevin, one final comment for something that I just noticed: Maybe we should rename this file to `release.yaml`, so that it would be consistent with te other SDKs.

kevihan-te replied:
Good idea. I updated. Thanks.

mpragosa-te replied:
I don't want to be picky... but it's still not fully consistent as we're using `yml` instead of `yaml` 😅
I know both are valid, but since we're trying to be consistent, I thought I should raise it. Sorry about that.

kevihan-te committed:
rename workflow filename

brumarqu-te commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,36 @@
+          major_version=$(echo "${{ inputs.releaseVersion }}" | cut -d '.' -f 1)
+          
+          sed -i "s|^\(var sdkVersion = \).*|\1\"${{ inputs.releaseVersion }}\"|" client/configuration.go
+          sed -i "s|^\(module github.com/thousandeyes/thousandeyes-sdk-go/\).*|\1$major_version|" go.mod
```
maybe we could use go itself to update this? Something like:
```
go mod edit -module github.com/thousandeyes/thousandeyes-sdk-go/v$major_version
```
The reason i'm asking to change, is to avoid messing up the `go.mod` file, ideally go itself should do it in the best way possible 😅 
This does adds the need for downloading go beforehand, so i'm happy to keep it as is if u feel strongly against.

brumarqu-te replied:
And btw there are other places where we'll need to update the version in case the major changes but it's mostly examples and README we can discuss it afterwards.

kevihan-te replied:
Good idea. Updated! Thanks Bruno.

brumarqu-te commented:
@kevihan-te can u check why the tests are failing?

kevihan-te committed:
use go mod edit to modify module version

kevihan-te committed:
use $GOPATH/bin/golint to call golint

kevihan-te committed:
install golint

kevihan-te commented:
> @kevihan-te can u check why the tests are failing?

Fixed!
